### PR TITLE
Fix Scaffold extend body

### LINKED ISSUE #157316: When using extendBody and the card & input box appear at the same time, the screen will flash white
### Steps to reproduce

When clicking on the input box to pop up the keyboard and then retract the keyboard, bottonNavigationBar will appear a white screen block flashing

### Expected results

fix

### Actual results

When clicking on the input box to pop up the keyboard and then retract the keyboard, bottonNavigationBar will appear a white screen block flashing

### Code sample

```
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(const MyApp());
  SystemChrome.setSystemUIOverlayStyle(const SystemUiOverlayStyle(
    statusBarColor: Colors.transparent,
    systemNavigationBarColor: Colors.transparent,
  ));
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return AnnotatedRegion<SystemUiOverlayStyle>(
      value: SystemUiOverlayStyle(
        systemNavigationBarColor: Colors.black.withOpacity(0),
      ),
      child: Scaffold(
        extendBody: true,
        appBar: AppBar(
          backgroundColor: Theme.of(context).colorScheme.inversePrimary,
          title: Text(widget.title),
        ),
        bottomNavigationBar: SafeArea(
          child: Container(
            decoration: BoxDecoration(
              color: Theme.of(context).colorScheme.surface,
              borderRadius: BorderRadius.circular(12),
              boxShadow: [
                BoxShadow(color: Theme.of(context).colorScheme.outlineVariant.withOpacity(0.5), blurRadius: 12),
              ],
            ),
            margin: const EdgeInsets.only(left: 38, right: 38, bottom: 7),
            padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 4),
            child: Row(
              children: [
                FilledButton(onPressed: () {}, child: const Text('add')),
                FilledButton(onPressed: () {}, child: const Text('add')),
              ],
            ),
          ),
        ),
        body: CustomScrollView(
          slivers: [
            const SliverToBoxAdapter(
              child: Column(
                children: [
                  ListTile(title: Text('Item 1')),
                  ListTile(title: Text('Item 2')),
                  ListTile(title: Text('Item 3')),
                  ListTile(title: Text('Item 4')),
                  ListTile(title: Text('Item 5')),
                  ListTile(title: Text('Item 6')),
                  ListTile(title: Text('Item 6')),
                  ListTile(title: Text('Item 6')),
                  ListTile(title: Text('Item 7')),
                  ListTile(title: Text('Item 8')),
                  ListTile(title: Text('Item 8')),
                ],
              ),
            ),
            const SliverToBoxAdapter(
              child: TextField(),
            ),
            SliverList.builder(
                itemBuilder: (v, a) {
                  return Card(
                    child: ListTile(title: Text('Item $a')),
                  );
                },
                itemCount: 10),
          ],
        ),
      ),
    );
  }
}
```

### Screenshots or Video

https://github.com/user-attachments/assets/7d503bd9-8ada-4622-9da4-8b9a8b96fad4

### Logs

<details open><summary>Logs</summary>

android

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

### PR DESCRIPTION
**Pull request by yiiim**
Fixes: #157316

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Review by nate-thegrate — COMMENTED** [review]
Thanks for fixing this!

File: packages/flutter/lib/src/material/scaffold.dart [hunk 1]
```diff
@@ -1086,9 +1086,9 @@ class _ScaffoldLayout extends MultiChildLayoutDelegate {
 
     if (hasChild(_ScaffoldSlot.body)) {
       double bodyMaxHeight = math.max(0.0, contentBottom - contentTop);
-      // extendBody locked when keyboard is open and the height is less than bottomWidgetsHeight.
-      final bool extendBody = this.extendBody && minInsets.bottom <= bottomWidgetsHeight;
-      if (extendBody) {
+      // When extendBody is true, the body is visible underneath the bottom widgets.
+      // This does not apply when the area is obscured by the device keyboard.
+      if (extendBody && minInsets.bottom <= bottomWidgetsHeight) {
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/scaffold.dart]: Ah, looks like the first part of my comment was used but not the `else` clause.

I'll probably go ahead and push a change here, feel free to revert the commit if you don't like it :)

File: packages/flutter/test/material/scaffold_test.dart [hunk 2]
```diff
@@ -3418,6 +3418,34 @@ void main() {
...
+          },
+        ),
+      );
+    }
+    await tester.pumpWidget(buildFrame());
+    expect(tester.getSize(find.byKey(bodyKey)).height, 600);
+    await tester.pumpWidget(buildFrame(keyboardHeight: 100));
+    expect(tester.getSize(find.byKey(bodyKey)).height, 600);
+    await tester.pumpWidget(buildFrame(keyboardHeight: 200));
+    expect(tester.getSize(find.byKey(bodyKey)).height, 400);
```
[the remark this answers is not in the document]
Comment (nate-thegrate) [reply packages/flutter/test/material/scaffold_test.dart]: Thanks for doing this!

**Review by nate-thegrate — APPROVED** [review]
LGTM, thanks again for the fix!

Feel free to land this one if current diffs look good to you :)

File: packages/flutter/lib/src/material/scaffold.dart (hunk 1 above)
Comment (yiiim) [reply packages/flutter/lib/src/material/scaffold.dart:1092]: I like this, I missed the code after the else before. :)

**Comment by yiiim** [comment]
What is making `Google testing` unhappy?

**Comment by nate-thegrate** [comment]
> What is making `Google testing` unhappy?

I'm not a Googler so idk, sorry 😔

Maybe you could see if a rebase fixes it? If not, we can get somebody else to step in.

**Comment by nate-thegrate** [comment]
success

woohoo!

**Comment by maxfrees** [comment]
@yiiim @nate-thegrate  when will it be released, or how can I use it in advance

**Comment by yiiim** [comment]
> @yiiim @nate-thegrate @Piinks when will it be released, or how can I use it in advance

Currently, it can only be used on the master channel. You can switch to the master channel using `flutter channel master`, but it is highly discouraged to do so as the master channel is usually very unstable.

**Pull request merged**